### PR TITLE
[PT-D] Multiprocessing API: standard streams: update 'to_std' utility parser to use enum membership testing

### LINKED ISSUE #92319: [PT-D]Track PT-D fixes for release blocker to upgrade to PT 3.11
**Overview**
Python 3.11 was [released in Oct 2022](https://www.python.org/downloads/release/python-3110/). It is the first release to include work from the “Faster CPython” workgroup. They advertise speedups between 10 and 60% depending on the workload without any change to the user code.

The issue for adding support in PyTorch [#86566](https://github.com/pytorch/pytorch/issues/86566) is already getting to the top of the list of issues when sorted by 👍 and 🚀reactions. We are also getting almost one issue a day at this point about people asking for conda support for 1.13 and/or because some features are broken in the current pip package we provide.
In the short term, with PyTorch 2.0 being a performance focused release, native python speedup can only be beneficial for the user experience.

Some of the main features of interest added here are:
Frozen imports / Static code objects (to speed up startup)
Cheaper, lazy Python frames
Inlined Python function calls
Specializing Adaptive Interpreter (a very basic jit)
More concise exception representation

Of course adding a new version of Python is a significant amount of work on the Infrastructure side (both for core and extended) and for this particular release, the low level changes in CPython lead to issues for PyTorch, especially for components like TorchScipt and Dynamo.

**The goal of this document is to have a full list of what needs to be done for 3.11 support to be on par with all the other python versions.
The goal is also to align on what we consider release blocker for 2.0 (if not everything can be done by then).**

Details for all distributed related failures are as follows:

**DISTRIBUTED** [#92301](https://github.com/pytorch/pytorch/issues/92301) 
Failures related to dataclasses
distributed/test_distributed_spawn for example
  File "/home/albandes/local/installs/python3.11/release/install/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'torch.distributed._shard.sharded_tensor.metadata.TensorProperties'> for field tensor_properties is not allowed: use default_factory
Most likely due to https://docs.python.org/3/whatsnew/3.11.html#dataclasses
One error test_launch_user_script
  File "/home/albandes/local/pytorch/3.11_release_source/torch/distributed/elastic/multiprocessing/api.py", line 162, in to_map
    map[i] = val_or_map.get(i, Std.NONE)
AttributeError: 'NoneType' object has no attribute 'get'
One error test_from_value
  File "/home/albandes/local/pytorch/3.11_release_source/test/distributed/elastic/multiprocessing/api_test.py", line 94, in test_from_value
    self.assertEqual(Std.NONE, Std.from_str("0"))

cc @ezyang @gchanan @zou3519 @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### PR DESCRIPTION
**Pull request by jayaddison**
May resolve #92319.

### CODE REVIEW

**Comment by jayaddison** [comment]
Would a [Developer Certificate of Origin](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) declaration be acceptable as an alternative to signing a Linux Foundation CLA?

**Review by fduwjj — APPROVED** [review]
LGTM and unblock for now. Please address the lint issue before merge, thanks!

File: torch/distributed/elastic/multiprocessing/api.py
Comment (wanchaol) [inline torch/distributed/elastic/multiprocessing/api.py:123]: can we raise a error in the else statement to silent mypy?

Comment (jayaddison) [reply torch/distributed/elastic/multiprocessing/api.py:123]: Could do.. hmm, thinking about it.  The intent here is not to change the behaviour of the code in any way.

Since the `else` should never be reached, then raising an error (like a `ValueError`) _should_ be fine.  In this case it might be preferable to use a `# type: ignore[return]` comment.

**Comment by jayaddison** [comment]
Thanks, @fduwjj @wanchaol!

**Comment by jayaddison** [comment]
(note: still some time required before possible merge, because I haven't signed the CLA.  I'd like to confirm whether a DCO is acceptable as an alternative, and have asked that question in #85559.  no problem for me if someone else solves the issue until then)

File: torch/distributed/elastic/multiprocessing/api.py
Comment (jayaddison) [reply torch/distributed/elastic/multiprocessing/api.py:123]: Also: was `mypy` not complaining about that before these changes?  That seems like a possible bug.

Comment (jayaddison) [reply torch/distributed/elastic/multiprocessing/api.py:123]: Eugh, perhaps it's more complicated.  At a guess: maybe `mypy` was able to infer from the `for` loop that all values are checked, and it doesn't do the same for the membership test.

Comment (jayaddison) [reply torch/distributed/elastic/multiprocessing/api.py:123]: :facepalm: Nope, a much simpler explanation :)

`mypy` did not complain previously because there were no type hints on the method.

d19f8f2f7385d8f501b481c5bcff15361e963f15 introduced type hints, and then the lint error appeared.

**Comment by fduwjj** [comment]
@jayaddison Have you tried to get the CLA directly as an individual contributor?

**Comment by jayaddison** [comment]
@fduwjj Not yet, nope - I'd prefer to sign with a Developer Certificate of Origin if possible (I admit there are pluses and minuses to each approach).

**Comment by jayaddison** [comment]
@fduwjj is this genuinely an emergency?  if so, I'll try to think of another way to get this fix applied (even, maybe, reluctantly signing the CLA) - however, if we can wait, then I'd prefer an answer about the DCO in #85559 (is there anyone we can ping about that?))

**Comment by fduwjj** [comment]
@jayaddison Maybe let's wait for one more week? Does that work for you? I am also asking around internally to see how to unblock you.

**Comment by jayaddison** [comment]
Sounds good to me - thanks @fduwjj (I'm also totally ok with anyone else implementing a similar fix)

**Comment by jayaddison** [comment]
For future reference, and transparency: although I [wasn't acknowledged by EasyCLA](https://github.com/pytorch/pytorch/pull/93209#issuecomment-1407453452) as having signed the PyTorch CLA, I _have_ previously signed a Facebook CLA (in the context of contributions to [`duckling`](https://github.com/facebook/duckling/)), and that's documented as [applicable to all of Facebook's open source projects](https://github.com/facebook/TextLayoutBuilder/blob/ec87ad1c811cbf82f73f4a1f33b470d52c115de7/.github/CONTRIBUTING.md#contributor-license-agreement-cla) -- [including PyTorch](https://github.com/pytorch/pytorch/blob/d9870d70c12dc59b0f8bce288910422bcb60b044/LICENSE#L3-L4) (there's no sublicensing for the `distributed` directory).

(probably sounds like I'm talking to myself, and to some extent I _am_ trying to understand the terms of my own contributions, because I think that's good practice for engineers, but also I know from experience that during -- hopefully rare and not always particularly enjoyable -- work involved in license/copyright checks, having clear information available can help reduce workloads; not too different to good reusable software - so maybe this comment can help, if it's ever required)

**Comment by jayaddison** [comment]
(and sorry @fduwjj, I didn't intend to hold this up unnecessarily - if I'd realized that my previous CLA signing was applicable at the time, I would've mentioned it.  it took a while for me to figure out the details)

**Pull request closed without merging**
